pkg/apis/core/v1alpha1/tiltrun_types.go

```diff
@@ -0,0 +1,274 @@
+
+	Items []TiltRun `json:"items"`
+}
+
+// TiltRunSpec defines the desired state of TiltRun
+type TiltRunSpec struct {
+	// TiltfilePath is the path to the Tiltfile for the run. It cannot be empty.
+	TiltfilePath string `json:"tiltfilePath"`
+	// ExitCondition defines the criteria for Tilt to exit.
+	ExitCondition ExitCondition `json:"exitCondition"`
+}
+
+type ExitCondition string
+
+const (
+	// ExitConditionManual cedes control to the user and will not exit based on resource status.
+	// This is used by `tilt up`.
+	ExitConditionManual ExitCondition = "manual"
+	// ExitConditionFirstFailure terminates upon the first encountered build or runtime failure.
+	// This is used by `tilt ci`.
+	ExitConditionFirstFailure ExitCondition = "first-failure"
+)
+
+var exitConditions = []ExitCondition{ExitConditionManual, ExitConditionFirstFailure}
+
+var _ resource.Object = &TiltRun{}
+var _ resourcestrategy.Validater = &TiltRun{}
+
+func (in *TiltRun) GetObjectMeta() *metav1.ObjectMeta {
+	return &in.ObjectMeta
+}
+
+func (in *TiltRun) NamespaceScoped() bool {
+	return false
+}
+
+func (in *TiltRun) New() runtime.Object {
+	return &TiltRun{}
+}
+
+func (in *TiltRun) NewList() runtime.Object {
+	return &TiltRunList{}
+}
+
+func (in *TiltRun) GetGroupVersionResource() schema.GroupVersionResource {
+	return schema.GroupVersionResource{
+		Group:    "tilt.dev",
+		Version:  "v1alpha1",
+		Resource: "tiltruns",
+	}
+}
+
+func (in *TiltRun) IsStorageVersion() bool {
+	return true
+}
+
+func (in *TiltRun) Validate(_ context.Context) field.ErrorList {
+	var fieldErrors field.ErrorList
+	if in.Spec.TiltfilePath == "" {
+		fieldErrors = append(fieldErrors, field.Required(field.NewPath("tiltfilePath"), "cannot be empty"))
+	}
+	validExitCondition := false
+	for _, v := range exitConditions {
+		if v == in.Spec.ExitCondition {
+			validExitCondition = true
+			break
+		}
+	}
+	if !validExitCondition {
+		var detailMsg strings.Builder
+		detailMsg.WriteString("valid values: ")
+		for i, ec := range exitConditions {
+			if i != 0 {
+				detailMsg.WriteString(", ")
+			}
+			detailMsg.WriteString(string(ec))
+		}
+
+		fieldErrors = append(fieldErrors, field.Invalid(
+			field.NewPath("exitCondition"),
+			in.Spec.ExitCondition,
+			detailMsg.String()))
+	}
+	return fieldErrors
+}
+
+var _ resource.ObjectList = &TiltRunList{}
+
+func (in *TiltRunList) GetListMeta() *metav1.ListMeta {
+	return &in.ListMeta
+}
+
+// TiltRunStatus defines the observed state of TiltRun
+type TiltRunStatus struct {
+	// PID is the process identifier for this instance of Tilt.
+	PID int64 `json:"pid"`
+	// StartTime is when the Tilt engine was first started and began processing resources.
+	StartTime metav1.MicroTime `json:"startTime"`
+	// Tiltfile is the current state of the Tilt engine including details about the last time the Tiltfile was
+	// processed and any build errors (e.g. syntax errors) encountered.
+	Tiltfile TiltfileState `json:"tiltfile"`
```

Comment:
A couple things that are kind of intertwined here:
* I kinda stuck with the way engine handles it where `Tiltfile` is special/top-level and not in the "user resources", but I dunno if that actually makes sense or if this should just be a known key in `Resources` slice?
* Going with the previous comment about not necessary _needing_ a `Tiltfile` in the future, this would then need to be `nil`-able and kind of awkward, which further leads me to thinking it should just be an entry in `Resources`

Reply:
One "strange" thing about mixing `Tiltfile` w/ the rest of the resources is that it's the exception to the "everything has runtime but not necessarily a build" as it's purely a build.

@nicks mentioned recently possibly getting rid of "build" as a concept (at this level), and I'm becoming more inclined to give that a shot. We could split resources (which is desirable from a DAG perspective anyway I think), so you would have `foo:build` and `foo:serve` resources.

Reply:
ya, i would be in favor of making the Tiltfile a generic resource, rather than something special (we initially made it special, but over time we've made it more like the other resources)

maybe it helps to think of Tiltfile as a Job that registers other resources on success?